crates/bevy_pbr/src/environment_map/filter.wgsl

```diff
@@ -0,0 +1,187 @@
+
+fn get_dir(u: f32, v: f32, face: u32) -> vec3<f32> {
+    switch face {
+        case 0u: { return vec3(1.0, v, -u); }
+        case 1u: { return vec3(-1.0, v, u); }
+        case 2u: { return vec3(u, 1.0, -v); }
+        case 3u: { return vec3(u, -1.0, v); }
+        case 4u: { return vec3(u, v, 1.0); }
+        default { return vec3(-u, v, -1.0); }
+    }
+}
+
+@compute
+@workgroup_size(64, 1, 1)
+fn main(@builtin(global_invocation_id) global_id: vec3<u32>) {
+    var id = global_id;
+    var level = 0u;
+    if id.x < 128u * 128u {
+        level = 0u;
+    } else if id.x < 128u * 128u + 64u * 64u {
+        level = 1u;
+        id.x -= 128u * 128u;
+    } else if id.x < 128u * 128u + 64u * 64u + 32u * 32u {
+        level = 2u;
+        id.x -= 128u * 128u + 64u * 64u;
+    } else if id.x < 128u * 128u + 64u * 64u + 32u * 32u + 16u * 16u {
+        level = 3u;
+        id.x -= 128u * 128u + 64u * 64u + 32u * 32u;
+    } else if id.x < 128u * 128u + 64u * 64u + 32u * 32u + 16u * 16u + 8u * 8u {
+        level = 4u;
+        id.x -= 128u * 128u + 64u * 64u + 32u * 32u + 16u * 16u;
+    } else if id.x < 128u * 128u + 64u * 64u + 32u * 32u + 16u * 16u + 8u * 8u + 4u * 4u {
+        level = 5u;
+        id.x -= 128u * 128u + 64u * 64u + 32u * 32u + 16u * 16u + 8u * 8u;
+    } else if id.x < 128u * 128u + 64u * 64u + 32u * 32u + 16u * 16u + 8u * 8u + 4u * 4u + 2u * 2u {
+        level = 6u;
+        id.x -= 128u * 128u + 64u * 64u + 32u * 32u + 16u * 16u + 8u * 8u + 4u * 4u;
+    } else {
+        return;
+    }
+
+    id.z = id.y;
+    let res = 128u >> level;
+    id.y = id.x / res;
+    id.x -= id.y * res;
+
+    let u = (f32(id.x) * 2.0 + 1.0) / f32(res) - 1.0;
+    let v = -(f32(id.y) * 2.0 + 1.0) / f32(res) + 1.0;
+
+    let dir = get_dir(u, v, id.z);
+    let frame_z = normalize(dir);
+    let adir = abs(dir);
+
+    var color = vec4(0.0);
+    for (var axis = 0u; axis < 3u; axis++) {
+        let other_axis0 = 1u - (axis & 1u) - (axis >> 1u);
+        let other_axis1 = 2u - (axis >> 1u);
+
+        let frame_weight = (max(adir[other_axis0], adir[other_axis1]) - 0.75) / 0.25;
+        if frame_weight > 0.0 {
+            var up_vector = vec3(0.0);
+            switch axis {
+                case 0u: { up_vector = vec3(1.0, 0.0, 0.0); }
+                case 1u: { up_vector = vec3(0.0, 1.0, 0.0); }
+                default { up_vector = vec3(0.0, 0.0, 1.0); }
+            }
+            let frame_x = normalize(cross(up_vector, frame_z));
+            let frame_y = cross(frame_z, frame_x);
+
+            var nx = dir[other_axis0];
+            var ny = dir[other_axis1];
+            let nz = adir[axis];
+
+            let nmax_xy = max(abs(ny), abs(nx));
+            nx /= nmax_xy;
+            ny /= nmax_xy;
+
+            var theta = 0.0;
+            if ny < nx {
+                if ny <= -0.999 { theta = nx; } else { theta = ny; }
+            } else {
+                if ny >= 0.999 { theta = -nx; } else { theta = -ny; }
```

Comment:
Kind of wonder if 0.999 should be factored out into an `EPSILON` constant.

Reply:
Eh, I'm just keeping it to match the original shader code